pantheon/src/main/java/tech/pegasys/pantheon/controller/PantheonControllerBuilder.java

```diff
@@ -0,0 +1,335 @@
+import tech.pegasys.pantheon.ethereum.blockcreation.MiningCoordinator;
+import tech.pegasys.pantheon.ethereum.chain.Blockchain;
+import tech.pegasys.pantheon.ethereum.chain.GenesisState;
+import tech.pegasys.pantheon.ethereum.chain.MutableBlockchain;
+import tech.pegasys.pantheon.ethereum.core.MiningParameters;
+import tech.pegasys.pantheon.ethereum.core.PrivacyParameters;
+import tech.pegasys.pantheon.ethereum.core.Synchronizer;
+import tech.pegasys.pantheon.ethereum.eth.EthProtocol;
+import tech.pegasys.pantheon.ethereum.eth.EthereumWireProtocolConfiguration;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthProtocolManager;
+import tech.pegasys.pantheon.ethereum.eth.peervalidation.DaoForkPeerValidator;
+import tech.pegasys.pantheon.ethereum.eth.peervalidation.PeerValidatorRunner;
+import tech.pegasys.pantheon.ethereum.eth.sync.DefaultSynchronizer;
+import tech.pegasys.pantheon.ethereum.eth.sync.SyncMode;
+import tech.pegasys.pantheon.ethereum.eth.sync.SynchronizerConfiguration;
+import tech.pegasys.pantheon.ethereum.eth.sync.state.SyncState;
+import tech.pegasys.pantheon.ethereum.eth.transactions.TransactionPool;
+import tech.pegasys.pantheon.ethereum.eth.transactions.TransactionPoolFactory;
+import tech.pegasys.pantheon.ethereum.jsonrpc.internal.methods.JsonRpcMethodFactory;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.ethereum.p2p.config.SubProtocolConfiguration;
+import tech.pegasys.pantheon.ethereum.storage.StorageProvider;
+import tech.pegasys.pantheon.ethereum.storage.keyvalue.RocksDbStorageProvider;
+import tech.pegasys.pantheon.ethereum.worldstate.WorldStateArchive;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+import tech.pegasys.pantheon.services.kvstore.RocksDbConfiguration;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.OptionalLong;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public abstract class PantheonControllerBuilder<C> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  protected GenesisConfigFile genesisConfig;
+  protected SynchronizerConfiguration syncConfig;
+  protected EthereumWireProtocolConfiguration ethereumWireProtocolConfiguration;
+  protected Integer networkId;
+  protected MiningParameters miningParameters;
+  protected MetricsSystem metricsSystem;
+  protected PrivacyParameters privacyParameters;
+  protected Path dataDirectory;
+  protected Clock clock;
+  protected Integer maxPendingTransactions;
+  protected KeyPair nodeKeys;
+  private StorageProvider storageProvider;
+  private final List<Runnable> shutdownActions = new ArrayList<>();
+  private RocksDbConfiguration rocksdDbConfiguration;
+
+  public static PantheonControllerBuilder<?> fromGenesisConfig(
```

Comment:
A builder with constructor parameters feels a bit off.  Perhaps these two `fromGenesisConfig` methods could be moved to a PantheonControllerBuilderFactory and this class could be AbstractPantheonControllerBuilder. The subclasses also don't add any builder functionality but are actually more of a strategy object for a generic factory.

Could this builder be a very simple builder that just holds the fields, and then in the build method delegates to the strategy/factory objects specific to the consensus types?  As late as possible?  Or perhaps keep the old class names and add a `.fromBuilder` method that will return an appropriate controller or throw an exception?

Reply:
Well it's not actually a constructor param - it's a static factory method to create the right type of builder based on the config.  I've moved it to a separate `PantheonControllerBuilderFactory` class since that's essentially what `PantheonCommand` wants. `PantheonCommandTest` wants to inject a mock `PantheonControllerBuilder` and was using a `Function<EthNetworkConfig, PantheonControllerBuilder>` but can now use a mock `PantheonControllerBuilderFactory`.

Whether it's a static method or a separate class the idea of having a builder factory displeases me greatly but I can't see any real way around it.

The key challenge in all of this is getting Java's generics to play nice. We need the genesis config before we can decide what generic type is required but if the builder isn't generic then it when it creates the `ProtocolSchedule` it has to declare it as `ProtocolSchedule<?>` but then it doesn't have the matching consensus context type the subclasses need.